fix(ottl): enforce valid types for setters in contexts (step 2)

jsvd opened the pull request:
#### Description

Setters in OTTL contexts now validate that values are of the expected type and return descriptive errors when type mismatches occur, improving debugging experience. This is the continuation of work done in #43505, addressing remaining context such as datapoint, profile, resource, span, and span event.

#### Link to tracking issue

Completes the fix to #40198, part 1 was #43505.

#### Testing

Tests were adapted to conform to the new error statements. New tests were not added given the simplicity of where ctxutil.ExpectType is applied to and given ctxutil.ExpectType itself was added with proper tests. Introducing tests for each setter would just be a large amount of code duplication, testing the exact same code pattern over and over.

#### Note for reviewer

This PR is likely best reviewed [without whitespace taken into consideration](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44881/changes?w=1) since most of the changes convert a block like

```go
			if v, ok := val.(Type); ok {
				// instruction
			}
```

into

```go
			v, err := ctxutil.ExpectType[Type](val)
			if err != nil {
				return err
			}
			// instruction
```

where any instruction or set of instructions will always be indented left after the refactor.

While it improves overall readability by putting the happy path logic outside a conditional, it makes reviewing the PR harder by touching a lot of code just for indentation.

jsvd committed:
fix(ottl): enforce valid types for setters in contexts (step 2)

Setters in OTTL contexts now validate that values are of the expected type and return
descriptive errors when type mismatches occur, improving debugging experience.
This is the continuation of work done in #43505, addressing remaining context such
as datapoint, profile, resource, span, and span event.

edmocosta commented:
> Tests were adapted to conform to the new error statements. New tests were not added given the simplicity of where ctxutil.ExpectType is applied to and given ctxutil.ExpectType itself was added with proper tests. Introducing tests for each setter would just be a large amount of code duplication, testing the exact same code pattern over and over.

The `ctxutil.ExpectType` tests does not guarantee the setters are working properly, and won't avoid removing the validation from the setters by mistake for example. We don't need to duplicate all the paths tests just for that, but it would be nice to have them covered as part of the existing suite, maybe something like this, enabling it for the changed paths, or disabling it for the ones not being validated yet?

```go
if tt.checkSetterType {
  err = accessor.Set(t.Context(), tCtx, struct{}{})
  require.ErrorContains(t, err, fmt.Sprintf("expects %T but got %T", tt.newVal, struct{}{}))
}
```

edmocosta commented on pkg/ottl/contexts/internal/ctxdatapoint/datapoint.go:
```diff
@@ -466,10 +488,12 @@ func accessBucketCounts[K Context]() ottl.StandardGetSetter[K] {
-				if histogramDataPoint, ok := tCtx.GetDataPoint().(pmetric.HistogramDataPoint); ok {
-					histogramDataPoint.BucketCounts().FromRaw(newBucketCount)
-				}
+			newBucketCount, err := ctxutil.ExpectType[[]uint64](val)
```
We can replace it with `ctxutil.SetCommonTypedSliceValues[uint64]`. The same for other slice paths.

jsvd committed:
address review comments

jsvd committed:
introduce default check in ctxutil/value.go's SetValue and remove error message checking

edmocosta commented on pkg/ottl/contexts/internal/ctxutil/value.go:
```diff
@@ -76,6 +76,8 @@ func SetValue(value pcommon.Value, val any) error {
 	case map[string]any:
 		err = value.FromRaw(v)
+	default:
+		return fmt.Errorf("unsupported type %T for SetValue", val)
```
```suggestion
		return fmt.Errorf("unsupported type %T for set operation; current value type is %v", val, value.Type())
```

Maybe not that helpful, but it might shed some light on the OTTL path we're referring to, especially for value updates.

jsvd committed:
add nil no-op guards to SetValue/SetMap and improve error message

Handle nil values as no-ops in SetValue and SetMap to preserve backward
compatibility. Include current value type in SetValue error message for
better debugging context. Add tests covering nil handling and error format.

jsvd committed:
Update .chloggen/ottl-setter-type-validation.yaml

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

evan-bradley reviewed (approved):
Great stuff. Thanks @jsvd!

evan-bradley committed:
Merge branch 'main' into fix_40198_part_2
